feat: wire up the write buffer to the ingester process with partition creation

pauldix opened the pull request:
This adds the scaffolding for the ingester server to consume data from Kafka. Future work will implement the actual buffering of data and add tests for the ingest handler.

This also includes the commit from #3531. Once that is merged in I'll rebase this branch to clean up the PR diff. In the meantime, viewing commit dda6563d475da90cf3f623c8b3f36e0f424c3142 will show what this PR is meant to add.

crepererum commented on ingester/src/handler.rs:
```diff
@@ -1,21 +1,61 @@
-    pub iox_catalog: Arc<dyn Catalog>,
+    /// Future that resolves when the background worker exits
+    #[allow(dead_code)]
+    join: Shared<BoxFuture<'static, Result<(), Arc<JoinError>>>>,
```
Can we not use a normal `JoinHandle` here? Then you could get away w/o the extra `CancellationToken` because `JoinHandle` has an `abort` method.

In either case though (using a join handle or this somewhat complicated construction here), I want you to stop the background thread when `IngestHandlerImpl` is dropped:

```rust
impl Drop for IngesterHandlerImpl {
    fn drop(&mut self) {
        // e.g.
        self.join.abort();
    }
}

pauldix replied:
Sounds good, I was just copying what was done elsewhere.

crepererum commented on ingester/src/handler.rs:
```diff
@@ -28,35 +68,161 @@ impl IngestHandlerImpl {
+        let dml_operation = match db_write_result {
+            Ok(db_write) => db_write,
+            // skip over invalid data in the write buffer so recovery can succeed
+            Err(e) => {
+                warn!(
+                    %e,
+                    %kafka_topic,
+                    %kafka_partition,
+                    "Error converting write buffer data to SequencedEntry",
+                );
+                continue;
```
I have seen similar patterns in my code and I think this is a bit hard to read. You're normally better off extracting the loop body into a function returning `Result<(), Error>` where you can just use `?` and then have a single `match` statement that ignores/logs the error.

BTW: I think we want some error count metric here.

crepererum commented on ingester/src/handler.rs:
```diff
@@ -28,35 +68,161 @@ impl IngestHandlerImpl {
-    pub fn get_topic_name(&self) -> String {
-        self.kafka_topic.name.clone()
-    }
+    while let Some(db_write_result) = stream.next().await {
```
Tracing is missing, but we (or I) can do that later. We are actually able to create tracing spans here that are linked to the traces on the router side, so we can correlate errors / weird behavior here all the way back to the frontend HTTP request that created the to-be-ingested data.

crepererum reviewed (commented):
Where are the tests? :wink:

domodwyer commented on influxdb_iox/src/commands/run/ingester.rs:
```diff
@@ -91,23 +103,60 @@ pub async fn command(config: Config) -> Result<()> {
-    };
+        .ok_or(Error::KafkaTopicNotFound(config.kafka_topic))?;
+
+    if config.kafka_partition_range_start > config.kafka_partition_range_end {
+        return Err(Error::KafkaRange);
+    }
```
Bit of a gotcha! The range below is:

```rust
(config.kafka_partition_range_start
        ..config.kafka_partition_range_end)
```

But the range end is not inclusive, so if you give it a "just read from sequencer 1" config:

```
kafka_partition_range_start=1
kafka_partition_range_end=1
```

You wind up with a range of `(1..1)` which [is of length 0](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=59a47f754bbdd3577f005175967cbf69).

It's probably less error prone to check that we have some sequencers once we're done initialising:

```rust
if kafka_partitions.is_empty() {
    return Err(...)
}
```

Might be good to document the range inclusiveness (or lack of!)

domodwyer reviewed (approved):
It's all coming together!

Might be worth making the `debug!(something semi-important has gone wrong)` into `WARN` level logs, so we see them in prod and act on them. For example, if we were never able to update the sequencer watermark we might not notice if it's logging at `DEBUG`.

pauldix commented on ingester/src/handler.rs:
```diff
@@ -34,54 +62,363 @@ impl IngestHandlerImpl {
+        );
+
+        let result = ingester_data
+            .buffer_operation(sequencer_id, dml_operation.clone())
```
This clone isn't ideal, but the metrics and span recorder are borrowing the DML operation. The data buffer actually needs to take ownership of the underlying mutable batches, which aren't used in the span and metrics at all, so there's probably metrics/span additions I can make so that they can work using the meta info, which isn't used in the buffer.

pauldix commented:
Ok, this is finally ready with the data going into the buffer. I moved a bunch of stuff over to `data.rs`. I need to add a bunch of tests there, but I think this is important to get in. The test in `handler.rs` goes through the entire pathway of getting data in the write buffer, reading it out, putting it into the data buffer and having the catalog get updated along the way.

I removed the `catalog_update.rs` for now because that is changing quite a bit and there wasn't anything really happening there yet. Next I'll be adding some basic lifecycle stuff which will have new hooks in `data.rs` to kick off persistence and then update the catalog after.

pauldix commented on ingester/src/data.rs:
```diff
@@ -47,70 +54,328 @@ pub enum Error {
+            _ => return Err(Error::TimeColumnNotPresent),
+        };
+
+        let partition_key = format!(
```
This bit will need to be fixed. To get things moving along, I just had it assume that all data within the mutable batch belongs to the same partition. I know this isn't yet guaranteed on the router side, but ideally it will be. I've logged #3603 to track that.

crepererum commented on ingester/src/handler.rs:
```diff
@@ -34,54 +62,363 @@ impl IngestHandlerImpl {
+                break;
+            }
+
+            assert!(t_0.elapsed() < Duration::from_secs(2));
+            tokio::time::sleep(Duration::from_millis(10)).await;
```
Pro tip for the future (you don't have to change that now): there's a tokio function for that, i.e. 

```rust
let t_0 = Instant::now();
loop {
    // do something

    assert!(t_0.elapsed() < Duration::from_secs(2));
    tokio::time::sleep(Duration::from_millis(10)).await;
}
```

can be written as

```rust
tokio::time::timeout(
    Duration::from_secs(2),
    async {
        loop {
            // do something

            tokio::time::sleep(Duration::from_millis(10)).await;
        }
    },
).await.expect("timeout");
```

Sure it's a bit more nested by the reader can immediately see that you're putting a timeout on a payload.

crepererum commented on dml/src/lib.rs:
```diff
@@ -258,6 +258,11 @@ impl DmlWrite {
     }
 
+    /// Moves the table names and mutable batches out of the map
+    pub fn tables_into_iter(self) -> impl Iterator<Item = (String, MutableBatch)> {
```
```suggestion
    pub fn into_tables(self) -> impl Iterator<Item = (String, MutableBatch)> {
```

The "Rusty" naming is `into_x` for transforming a container into an iterator of it's parts. See `{HashMap,BTreeMap}::{into_keys,into_values}` for example.

crepererum reviewed (changes requested):
Found some code quality issues and minor bugs, but overall I think this works.

domodwyer commented on ingester/src/data.rs:
```diff
@@ -47,70 +54,328 @@ pub enum Error {
+                self.buffer_delete(
+                    delete
+                        .table_name()
+                        .expect("table name must exist in delete"),
```
The router does not resolve it in any way - it just passes the delete predicate into the DML op and pushes it down the tube to the ingester.

I'm not sure we could resolve it in the router without it being racy with concurrent table creation in another router - worth a discussion 👍

pauldix replied:
I thought that this was going to get enforced at the HTTP API level to require a table. @e-dard, I believe you looked into the behavior on the delete API?

crepererum replied:
I'm pretty sure the HTTP API allows that because we deal w/ that in the current ingester code and @tustvold and I thought about it multiple times when pondering about delete routing.

tustvold replied:
Deletes without tables I'm pretty certain are allowed, the thing we decided not to support is deletes with `__field` predicates as they aren't really deletes but upserts.

One thing to perhaps bear in mind, is that it isn't always trivial to infer what tables a delete concerns. This is something that has come up in the pruning logic, you can stick `__measurement` anywhere including in a nested chain of expressions. The current ingest tier is pessimistic in this regard

pauldix commented on ingester/src/data.rs:
```diff
@@ -47,70 +54,328 @@ pub enum Error {
+        let min_time = Timestamp::new(predicate.range.start());
+        let max_time = Timestamp::new(predicate.range.end());
+
+        let tombstone = catalog
+            .tombstones()
+            .create_or_get(
+                table_data.table_id,
+                sequencer_id,
+                sequence_number,
+                min_time,
+                max_time,
+                &predicate.expr_sql_string(),
+            )
+            .await
+            .context(CatalogSnafu)?;
```
I don't think we'll need the optimization for some time as we don't have that many deletes hitting any single cluster.

pauldix commented on ingester/src/handler.rs:
```diff
@@ -34,54 +62,363 @@ impl IngestHandlerImpl {
+        let kafka_topic_name = topic.name.clone();
+        let ingest_metrics = WriteBufferIngestMetrics::new(registry, &topic.name);
+
+        let write_buffer: &'static mut _ = Box::leak(write_buffer);
```
Ah yes, I forgot to add a comment on this! Thanks for adding the context 😄

pauldix commented on ingester/src/handler.rs:
```diff
@@ -34,54 +62,363 @@ impl IngestHandlerImpl {
+    use time::Time;
+    use write_buffer::mock::{MockBufferForReading, MockBufferSharedState};
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
```
Ah, I added this before because the test was freezing up, but I just took it out and everything completes correctly. Will remove.

pauldix commented on ingester/src/data.rs:
```diff
@@ -47,70 +54,328 @@ pub enum Error {
+        table_name: &str,
+        catalog: &dyn Catalog,
+    ) -> Result<Arc<TableData>> {
+        let table = catalog
+            .tables()
+            .create_or_get(table_name, self.namespace_id)
+            .await
+            .context(CatalogSnafu)?;
+        let mut t = self.tables.write();
+        let data = Arc::new(TableData::new(table.id));
+        t.insert(table.name, Arc::clone(&data));
```
Good catch!

pauldix commented on ingester/src/handler.rs:
```diff
@@ -1,27 +1,55 @@
 //! Ingest handler
```
yeah, not really a handler. More like the running ingester server. Should I rename in a followup PR?

pauldix commented on ingester/src/data.rs:
```diff
@@ -1,16 +1,22 @@
 //! Data for the lifecycle of the Ingester
```
haha good point. I'll add it in a followup PR 😄

pauldix commented on ingester/src/data.rs:
```diff
@@ -47,70 +54,328 @@ pub enum Error {
+        let sequence_number = i64::try_from(sequence_number).expect("sequence out of bounds");
+        let sequence_number = SequenceNumber::new(sequence_number);
+
+        match dml_operation {
```
The operation actually has to be taken apart in the namespace because a write can apply to many tables.

crepererum reviewed (approved):
OK, let's merge this and let's improve docs / naming in a follow-up PR, so we can unblock the team and get a test bench run.
